core/ir/synthetic/encoding_common.h

```diff
@@ -33,8 +33,84 @@
+ * host is an x86 architecture.
+ * Also note that in DynamoRIO we use the #instr_t ISA mode to determine what type of
+ * encoding to perform, and #dr_context_t ISA mode to determine what type of decoding to
+ * perform.
+ * Currently the only exception to this rule happens for decoding of synthetic
+ * instructions, where #instr_t ISA mode takes precedence and #dr_context_t ISA mode is
+ * ignored if #instr_t ISA mode is DR_ISA_SYNTHETIC.
+ * XXX i#1684: Note that this is part of a larger issue, where lack of cross-arch support
+ * in the same build of DynamoRIO is limiting us.
+ */
+
+/* Here we describe the encoding scheme for the Synthetic ISA that is enforced in decode.c
+ * and encode.c.
+ *
+ * Encoded instructions are 4 bytes aligned.
+ *
+ * All instruction encodings begin with the following 4 bytes, which follow this scheme:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *       category          eflags #src   #dst
+ *
+ * 22 bits, category: it's a high level representation of the opcode of an instruction.
+ * Each bit represents one category following #dr_instr_category_t.
+ * Note that an instruction can belong to more than one category, hence multiple bits can
+ * be set.
+ * 2 bits, eflags: most significant bit set to 1 indicates the instruction reads at least
+ * one arithmetic flag, least significant bit set to 1 indicates the instruction writes
+ * at least one arithmetic flag.
+ * 4 bits, #src: number of source operands (read) that are registers.
+ * 4 bits, #dst: number of destination operands (written) that are registers.
+ * (Note that we are only interested in register dependencies, hence operands that are
+ * not registers, such as immediates or memory references, are not present.)
+ *
+ * Following the 4 instruction-related bytes are the bytes for encoding register operands.
+ * Each operand is 1 byte.
+ * The destination operands go first, followed by the source operands.
+ * An instruction can have up to 8 operands (sources + destinations).
+ * Note that, because of 4 bytes alignment, instructions with 1 to 4 (included) operands
+ * will have a size of 8 bytes (4 instruction-related bytes + 4 operand-related bytes),
+ * while instructions with 5 to 8 (included) operands will have a size of 12 bytes
+ * (4 instruction-related bytes + 8 operand-related bytes).
+ * Instructions with no operands only have 4 bytes (the 4 instruction-related bytes).
+ * For example, an instruction with 3 operands (1 dst, 2 src) has 4 additional bytes that
+ * are encoded following this scheme:
+ * |--------| |--------| |--------| |--------|
+ * 31.. ..24  23.. ..16  15..  ..8  7..   ..0
+ *            src_opnd1  src_opnd0  dst_opnd0
+ *
+ * Because of 4 bytes alignment, the last byte (31..  ..24) is padding and it's undefined
+ * (i.e., it cannot be assumed that it has been zeroed-out or contains any meaningful
+ * value).
+ *
+ * We assume all encoded values to be little-endian.
```

Comment:
Well, the operands are now a byte stream, so there is no endian-ness concept there; the 4-byte header word is little-endian as shown with the bit layout.  So not sure we need this sentence: it may be confusing?

Reply:
Oh, yes, that was referring to the header bytes, moved there.